lib/Doctrine/DBAL/Connections/PrimaryReplicaConnection.php

```diff
@@ -0,0 +1,361 @@
+        if (! isset($params['replica'], $params['primary'])) {
+            throw new InvalidArgumentException('primary or replica configuration missing');
+        }
+
+        if (count($params['replica']) === 0) {
+            throw new InvalidArgumentException('You have to configure at least one replica.');
+        }
+
+        $params['primary']['driver'] = $params['driver'];
+        foreach ($params['replica'] as $replicaKey => $replica) {
+            $params['replica'][$replicaKey]['driver'] = $params['driver'];
+        }
+
+
+        parent::__construct($params, $driver, $config, $eventManager);
+    }
+
+    private function switchToPrimary(): bool
+    {
+        $forcePrimaryAsReplica = $this->getTransactionNestingLevel() > 0;
+
+        if (isset($this->primary)) {
+            $this->_conn = $this->primary;
+
+            if ($forcePrimaryAsReplica) {
+                $this->replica = $this->_conn;
+            }
+
+            return false;
+        }
+
+        $this->primary = $this->_conn = $this->connectTo('primary');
+
+        // Set replica connection to primary to avoid invalid reads
+        $this->replica = $this->primary;
+
+        if ($this->_eventManager->hasListeners(Events::postConnect)) {
+            $eventArgs = new ConnectionEventArgs($this);
+            $this->_eventManager->dispatchEvent(Events::postConnect, $eventArgs);
+        }
+
+        return true;
+    }
+
+    public function connect(): bool
+    {
+        $connectionName = 'replica';
+
+        if ($this->_conn !== null) {
+            return false;
+        }
+
+        $forcePrimaryAsReplica = false;
+
+        if ($this->getTransactionNestingLevel() > 0) {
+            $connectionName        = 'primary';
+            $forcePrimaryAsReplica = true;
+        }
+
+        if (isset($this->$connectionName)) {
+            $this->_conn = $this->$connectionName;
+
+            if ($forcePrimaryAsReplica) {
+                $this->replica = $this->_conn;
+            }
+
+            return false;
+        }
+
+        if ($connectionName === 'primary') {
+            $this->primary = $this->_conn = $this->connectTo($connectionName);
+
+            // Set replica connection to primary to avoid invalid reads
+            $this->replica = $this->primary;
+        } else {
+            $this->replica = $this->_conn = $this->connectTo($connectionName);
+        }
+
+        if ($this->_eventManager->hasListeners(Events::postConnect)) {
+            $eventArgs = new ConnectionEventArgs($this);
+            $this->_eventManager->dispatchEvent(Events::postConnect, $eventArgs);
+        }
+
+        return true;
+    }
+
+    /**
+     * Connects to a specific connection.
+     */
+    private function connectTo(string $connectionName): DriverConnection
+    {
+        $params = $this->getParams();
+
+        $driverOptions = $params['driverOptions'] ?? [];
+
+        $connectionParams = $this->chooseConnectionConfiguration($connectionName, $params);
```

Comment:
So does this mean that every time we "switch" between the primary and replica endpoints, the replica will be randomly chosen and the old connection will be closed? What's the rationale behind this behavior?

Reply:
:thinking: we are not supposed to switch from the primary to the replica, are we?

Reply:
This is really questionable behavior. If I insert a "user logged in" event in one table and then query a heavy report from the other, why should the report use the master connection?

Reply:
The step from primary to replica is not really a common use-case, but what you describe could be an example. If the application does a single write early in the request and you can isolate that and don't need to care about replication lag with this write, then you could switch back to a replica. The idea would be:

```php
$connection->connect('primary');
$connection->executeUpdate("....");
$connection->connect("replica");
```

Reply:
Why not just do `$primaryConnection->executeUpdate(); $replicaConnection->doWhatever();`? Why keep one connection as a shared state and switch between them?

Reply:
The answer is probably "convenience" (at the cost of some usefulness). Like always with such things ;)